core/Binary.carp

```diff
@@ -0,0 +1,363 @@
+;; TODO: Replace all of these with a single type-generic
+;; zip-n macro.
+(defmodule Maybe
+  (defn zip [f a b]
+    (match a
+    (Maybe.Nothing) (Maybe.Nothing)
+    (Just x)
+      (match b
+      (Maybe.Nothing) (Maybe.Nothing)
+      (Just y) (Just (~f x y)))))
+
+  (defn zip4 [f a b c d]
+    (match a
+    (Maybe.Nothing) (Maybe.Nothing)
+    (Just x)
+      (match b
+      (Maybe.Nothing) (Maybe.Nothing)
+      (Just y)
+        (match c
+        (Maybe.Nothing) (Maybe.Nothing)
+        (Just z)
+          (match d
+          (Maybe.Nothing) (Maybe.Nothing)
+          (Just x2) (Just (~f x y z x2)))))))
+
+  (defn zip8 [f a b c d e a1 b1 c1]
+   (match a
+   (Maybe.Nothing) (Maybe.Nothing)
+   (Just x)
+     (match b
+     (Maybe.Nothing) (Maybe.Nothing)
+     (Just y)
+       (match c
+       (Maybe.Nothing) (Maybe.Nothing)
+       (Just z)
+         (match d
+         (Maybe.Nothing) (Maybe.Nothing)
+         (Just x2)
+           (match e
+           (Maybe.Nothing) (Maybe.Nothing)
+           (Just y2)
+             (match a1
+             (Maybe.Nothing) (Maybe.Nothing)
+             (Just z2)
+               (match b1
+               (Maybe.Nothing) (Maybe.Nothing)
+               (Just x3)
+                 (match c1
+                 (Maybe.Nothing) (Maybe.Nothing)
+                 (Just y3) (Just (~f x y z x2 y2 z2 x3 y3)))))))))))
+)
+
+(defmodule Binary
+  (doc Order
+    "The type of byte orders.
+
+    LittleEndian designates the little endian ordering, and indicates the least
+    significant byte appears first in a given byte sequence.
+
+    BigEndian designates the big endian ordering, and indicates the most
+    significant byte occurs first in a given byte sequence.")
+  (deftype Order LittleEndian BigEndian)
+
+  (register to-int16 (λ [Byte Byte] Uint16))
+  (register to-int32 (λ [Byte Byte Byte Byte] Uint32))
+  (register to-int64 (λ [Byte Byte Byte Byte Byte Byte Byte Byte] Uint64))
+  (register int16-to-byte  (λ [(Ref Uint16)] Byte))
+  (register int32-to-byte  (λ [(Ref Uint32)] Byte))
+  (register int64-to-byte  (λ [(Ref Uint64)] Byte))
+  (register system-endianness-internal (λ [] Int))
+
+  (defn unwrap-success [x]
+    (Result.unwrap-or-zero @x))
+
+  (defn unwrap-error [x]
+    (Result.from-error @x (zero)))
+
+  (doc byte-converter
+    "Returns a function that, when called, attempts to convert an array of bytes using `f` and `order`
+
+     If the conversion is successful, returns a `Result.Success` containing the converted value.
+
+     If the conversion fails, returns a `Result.Error` containing the byte array passed as an argument.")
+  (defn byte-converter [f order]
+     (fn [bs]
+       (match (~f order bs)
+              (Maybe.Nothing) (Result.Error @bs)
+              (Maybe.Just i)  (Result.Success i))))
+  (doc interpreted
+    "Returns the interpreted value from a sequence of byte-converion results")
+  (private interpreted)
+  (defn interpreted [results]
```

Comment:
Ahh yes! good ol' fusion! Thanks for the tip!

Actually, it'd be really sweet if the compiler did automatic fusion for us...is that something you're interested in investigating :) I have several papers on fusion that I could pass along (not sure precisely whether or not any of them discuss actually implementing it in a compiler though)

Reply:
Soooo, I'm going to hold off on this until a later PR. There seems to be a slight bug in reduce whereby it's interpreting the function argument array as a literal array I'm getting:

```
Arrays can’t contain references: `x` at line 94, column 15 in '/Users/scottolsen/dev/Carp/core/Binary.carp'.

You’ll have to make a copy using `@`.
```

With the definition
```
(defn interpreted [results] 
  (let [f (fn [x y] (if (Result.success? y)
                                (Array.push-back x y)
                                x)]
    (Array.reduce &f [] results))
```

Where the function definition in the let is at line 94. There's also a good chance I'm just writing this incorrectly.

Reply:
I think the problem is the way you use `reduce` and `push-back` in combination.

Their respective type signatures are:
```
Array.reduce : (λ [(Ref (λ [a, (Ref b c)] a d) e), a, (Ref (Array b) c)] a)
Array.push-back : (λ [(Array a), a] (Array a))
```

The `reduce` signature is a bit hard to read, but the main problem is the function that is passed in, so let’s zoom in on that:

```
(λ [a, (Ref b c)] a d)
```

Let’s remove the lifetime params, because they look a little messy:

```
(λ [a, (Ref b)] a)
```

The array element is passed in by reference! That makes sense, `Result.success?` takes a reference. But let’s look at `Array.push-back` again. It doesn’t take a reference, it takes the argument! Thus, `x` is inferred to be a `(Array (Ref (Result a b)))` (because we push references in), which is invalid!

Those are tricky to figure out, but basically you’ll have to add a copy in `push-back`, like so: `(Array.push-back x @y)`.

Reply:
Thanks for the detailed explanation! I was trying a couple of different variants, but didn't think of this one. I'll give it a shot!

Maybe we can better distinguish lifetimes from type params? I also find types that have some combination of both difficult to read at times.

Reply:
Yeah, any ideas for how to make them stand out (like the tick in Rust) would be greatly appreciated. I just haven't come up with any good ideas...

Reply:
It's tricky! 

I kind of like alligator brackets, it takes up a little extra space, but it makes them really obvious:

`(λ [(Ref (λ [a, (Ref b <c>)] a <d>) <e>), a, (Ref (Array b) <c>)] a)`

I think `{}` could be nice too, but it's harder to read than `<>`:

`(λ [(Ref (λ [a, (Ref b {c})] a {d}) {e}), a, (Ref (Array b) {c})] a)`

For single character options, I think stuff that's equally obvious is best:

`(λ [(Ref (λ [a, (Ref b *c)] a *d) *e), a, (Ref (Array b) *c)] a)`
`(λ [(Ref (λ [a, (Ref b #c)] a #d) #e), a, (Ref (Array b) #c)] a)`

Or you could overload the `&` or `@` which would make it clear that it's specific to references:

`(λ [(Ref (λ [a, (Ref b &c)] a &d) &e), a, (Ref (Array b) &c)] a)`
`(λ [(Ref (λ [a, (Ref b @c)] a @d) @e), a, (Ref (Array b) @c)] a)`

But that could be confusing, too.

My personal preference is for `<a>`-- not sure if we're using that to mean something else already, though.

Reply:
Yeah, alligator brackets (I always called them angle brackets, but this name is way cooler, thanks @scolsen) look good to me!